discovery/spec.md

```diff
@@ -820,8 +820,8 @@ exist) then the entire request MUST fail with no effect.
 If an `id` is not specified in any of the Services the entire request MUST
 fail.  If no Service can be found that corresponds to a specified `id`, the
 entire request MUST fail.  An `epoch` value MAY be omitted.  If a Service
-corresponding to a specified `id` has a higher recorded `epoch` value than the
-one given the entire request MUST fail.
+corresponding to a specified `id` has a different recorded `epoch` value than
```

Comment:
This changes the behavior that the client cannot supply the value of an epoch.  This may be fine, just being explicit.  Consider the case where the epoch is a sequence ID or monotonic offset from a log.

Reply:
does the value need to be higher to retain the semantics we intended? I was assuming that the purpose of including the epoch was to ensure that the client was issuing the delete() on the version they think is the latest. And if the latest.epoch != client.epoch then they wanted the delete to fail - regardless of whether the client's is old (or even newer). Is that not your understanding?

Reply:
Epoch is defined [earlier in the doc](https://github.com/cloudevents/spec/blob/b6f972af0502cd3d38291e63ee3ab9ae995e6a22/discovery/spec.md#epoch):

> For example, implementations are free to use a value that represents a timestamp or could choose to simply use a monotonically increasing number. The only requirement is that the value MUST always increase each time the Service Entry is updated.

The positive monotonicity of the epoch has been part of the definition for some time.  The question is something like whether the semantic of the epoch given is "here's the new epoch to use" or "here's the epoch I am targeting".  #928 seems to also have changed this semantic.  If you look under POST the convention that has been previously followed across these management APIs was that the caller was supplying the value for the epoch the record update was occurring under.

Reply:
If we are going to change it, perhaps we do this as  part of a separate PR and across the endpoints?

Reply:
> The question is something like whether the semantic of the epoch given is "here's the new epoch to use" or "here's the epoch I am targeting"

but we're talking about delete() here, not update.

Let me ask this... what do you consider to be the purpose behind providing an epoch on delete() ? By providing it the client is asking for some kind of check, otherwise they wouldn't include it at all and the service would always be deleted.
So, It seems like there are 3 possible conditions:
1 - provided epoch is less than the DE service's epoch's value. Which means the client had an old version. And the only possible result here is to reject the request because it's stale. If the user wanted some other result to happen then it shouldn't have included the epoch at all.
2 - provided epoch is the same as the DE service's epoch value. The result here must be to delete it.
3 - provided epoch is higher than the DE service's epoch value. Technically this should never happen unless they just made up a value, which is probably leaving the scope of "proper usage of the spec". But if we had to choose an action, I'd say we should reject the delete() to be on the safe side.

So, if we really only need to worry about 1 and 2, then we're really only talking about "are the epoch values different" - aren't we?

Reply:
> but we're talking about delete() here, not update.

Yes.  Have we stopped supporting soft deletes?

Reply:
> 3 - provided epoch is higher than the DE service's epoch value. Technically this should never happen unless they just made up a value, which is probably leaving the scope of "proper usage of the spec"

It's not clear to me that using the offset of an event on a particular partition of a distributed log as the epoch is "made up" or an improper use of the service.  In fact, I can imagine a fair bit of utility to using that value.

Reply:
> Have we stopped supporting soft deletes?

We don't have soft deletes in the spec right now. We've talked about them but I don't think we ever added them.

Reply:
You are right that we don't discuss soft deletes in the spec.  I [originally included but then removed it](https://github.com/cloudevents/spec/pull/837/commits/3846c460bba31f64e33098f296df026bc75bed69) when I was cleaning up this API in preparation for the history PR.  As I recall the discussion we decided that by not mentioning it we were not prohibiting (i.e. allowing) the practice without having to get into the complexity of its ramifications of specifying the behavior.

Reply:
I prefer the old wording.

Exact epoch match seems like a larger burden for the client, I think there is a valid case for the client having a more up to date catalog than the server and wanting to delete the server's copy to update? This way requires a refetch and is racy.

Reply:
@n3wscott can you elaborate on the scenario? When would a client have a higher value? And even if they did, and if they didn't care, wouldn't they just leave off the epoch all together?

Reply:
Scott can speak for himself but replication seems like a scenario where the client replicating the later version of the record necessarily holds a higher epoch (because the replication target has last version before receiving the update).

Reply:
yup that scenario crossed my mind too. However, it seems to me that if we want to ensure that we support replication of Discovery Endpoints then we have a lot more stuff to do beyond DELETE. As of now the spec doesn't address it at all so I'm more focused on the "this DE is the single source of truth" scenario, and in that case a client with a higher epoch is more like an error.

I will say however that during our other epoch discussion we might quickly get into the replication scenario - I just didn't want to jump the gun or do a partial solution.